### PR TITLE
[feature](nereids)add InPredicate in expressions

### PR DESCRIPTION
**Pull request by YangShaw**
# Proposed changes

## Problem Summary:
- Add InPredicate expression parser and translator
- Add regression-test for In predicate (in nereids_syntax)
- Support NOT EqualTo and NOT InPredicate in ExpressionTranslator#visitNot()
## Checklist(Required)

1. Type of your changes: (Feature)
2. Does it affect the original behavior: (No)
3. Has unit tests been added: (Yes)
4. Has document been added or modified: (No Need)
5. Does it need to update dependencies: (No)
6. Are there any changes that cannot be rolled back: (No)
## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -783,7 +784,10 @@ private Expression withPredicate(Expression valueExpression, PredicateContext ct
                 case DorisParser.IN:
                     if (ctx.query() == null) {
                         //TODO: InPredicate
```
Comment (zhengshiJ) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:786]: Comments can be deleted

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * In predicate expression.
+ */
+public class InPredicate extends Expression {
+
+    private Expression compareExpr;
+    private List<Expression> optionsList;
+
+    public InPredicate(Expression compareExpr, List<Expression> optionsList) {
+        super(new Builder<Expression>().add(compareExpr).addAll(optionsList).build().toArray(new Expression[0]));
+        this.compareExpr = compareExpr;
```
Comment (zhengshiJ) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java:42]: You can use Objects.requireNonNull to ensure that the parameter cannot be null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 1 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java (hunk 2 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: Has added requireNonNull() for compareExpr

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Override
+    public DataType getDataType() throws UnboundException {
+        return BooleanType.INSTANCE;
+    }
+
+    @Override
+    public boolean nullable() throws UnboundException {
+        return optionsList.stream().map(Expression::nullable)
+            .reduce((a, b) -> a || b).get();
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java:58]: InPredicate is nullable if compareExpr is nullable?
```suggestion
        return children().stream().anyMatch(Expression::nullable);
```

Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: Yes, I missed it. Have fixed.

**Review by zhengshiJ — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            .collect(Collectors.joining(",", "(", ")"));
+    }
+
+    @Override
+    public String toSql() {
+        return compareExpr.toSql() + " IN " + optionsList.stream()
+            .map(Expression::toSql)
+            .collect(Collectors.joining(",", "(", ")"));
+    }
+
```
[the remark this answers is not in the document]
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java:73]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 5]
```diff
@@ -838,4 +840,10 @@ public Expression visitSubqueryExpression(SubqueryExpressionContext subqueryExpr
     public Expression visitExist(ExistContext context) {
         return ParserUtils.withOrigin(context, () -> new Exists(typedVisit(context.query())));
     }
+
+    public List<Expression> withInList(PredicateContext ctx) {
+        List<Expression> expressions = ctx.expression().stream()
+                .map(expr -> getExpression(expr)).collect(ImmutableList.toImmutableList());
+        return expressions;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:847]:
```suggestion
        return ctx.expression().stream()
                .map(this::getExpression).collect(ImmutableList.toImmutableList());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * In predicate expression.
+ */
+public class InPredicate extends Expression {
+
+    private Expression compareExpr;
+    private List<Expression> optionsList;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java:38]: final

Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java:38]:
```suggestion
    private List<Expression> options;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 7]
```diff
@@ -798,9 +799,10 @@ private Expression withPredicate(Expression valueExpression, PredicateContext ct
                     break;
                 case DorisParser.IN:
                     if (ctx.query() == null) {
-                        //TODO: InPredicate
-                        outExpression = null;
-                        throw new IllegalStateException("Unsupported predicate type: " + ctx.kind.getText());
+                        outExpression = new InPredicate(
+                            valueExpression,
+                            withInList(ctx)
```
[the remark this answers is not in the document]
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java (hunk 5 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java (hunk 6 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * In predicate expression.
+ */
+public class InPredicate extends Expression {
+
+    private Expression compareExpr;
```
[the remark this answers is not in the document]
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    @Override
+    public boolean nullable() throws UnboundException {
+        return children().stream().anyMatch(Expression::nullable);
+    }
+
+    @Override
+    public String toString() {
+        return compareExpr + " IN " + optionsList.stream()
+            .map(Expression::toString)
+            .collect(Collectors.joining(",", "(", ")"));
```
[the remark this answers is not in the document]
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    public String toSql() {
+        return compareExpr.toSql() + " IN " + optionsList.stream()
+            .map(Expression::toSql)
+            .collect(Collectors.joining(",", "(", ")"));
```
[the remark this answers is not in the document]
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java (hunk 6 above)
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/InPredicate.java]: done

**Review by morrySnow — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java [hunk 11]
```diff
@@ -141,10 +145,48 @@ public Expr visitNullSafeEqual(NullSafeEqual nullSafeEqual, PlanTranslatorContex
...
+                    likePredicate,
+                    null);
+            return compoundPredicate;
+        } else if (not.child() instanceof Regexp) {
+            Regexp regexp = (Regexp) not.child();
+            LikePredicate likePredicate = (LikePredicate) visitRegexp(regexp, context);
+            CompoundPredicate compoundPredicate = new CompoundPredicate(CompoundPredicate.Operator.NOT,
+                    likePredicate,
+                    null);
+            return compoundPredicate;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java:170]: could merge into else？

Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java:170]: I will have a try.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java [hunk 12]
```diff
@@ -141,10 +145,48 @@ public Expr visitNullSafeEqual(NullSafeEqual nullSafeEqual, PlanTranslatorContex
...
+        } else if (not.child() instanceof Between) {
+            Between between = (Between) not.child();
+            BetweenPredicate betweenPredicate = new BetweenPredicate(
+                    between.getCompareExpr().accept(this, context),
+                    between.getLowerBound().accept(this, context),
+                    between.getUpperBound().accept(this, context),
+                    true);
+            return betweenPredicate;
```
[the remark this answers is not in the document]
Comment (YangShaw) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java:178]: Yes, I have realized this after ran some queries. I will delete this case and reset visitBetween later.

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**